### PR TITLE
Store TLS version in SSL session structure

### LINKED ISSUE #4802: Prepare SSL session serialization for introduction of TLS 1.3 session (tickets)
Currently, session serialization in TLS 1.2 does not store an identifier for the TLS-version used by the session, and uses only a fixed (for a given config) serialization layout.

With the introduction of support for TLS 1.3, we will need a field indicating whether the serialized session is for TLS 1.2 or TLS 1.3. Moreover, for TLS 1.3, we will want to use a different - more compact - format for the serialized data which will look something like this:

```c
/*
 * Serialize a session in the following format:
 * (in the presentation language of TLS, RFC 8446 section 3)
 *
 *  struct {
 *
 *    opaque mbedtls_version[3];   // major, minor, patch
 *    opaque session_format[2];    // version-specific 16-bit field determining
 *                                 // the format of the remaining
 *                                 // serialized data.
 *  
 *          Note: When updating the format, remember to keep
 *          these version+format bytes.
 *  
 *                                 // In this version, `session_format` determines
 *                                 // the setting of those compile-time
 *                                 // configuration options which influence
 *                                 // the structure of mbedtls_ssl_session.
 *  
 *    uint8_t minor_ver;           // Possible values:
 *                                 // - TLS 1.2 (MBEDTLS_SSL_MINOR_VERSION_3) 
 *                                 // - TLS 1.3 (MBEDTLS_SSL_MINOR_VERSION_4) 
 *   
 *    select (serialized_session.minor_ver) {
 *
 *      case MBEDTLS_SSL_MINOR_VERSION_3: // TLS 1.2
 *        uint64 start_time;
 *        uint8 ciphersuite[2];           // defined by the standard
 *        uint8 compression;              // 0 or 1
 *        uint8 session_id_len;           // at most 32
 *        opaque session_id[32];
 *        opaque master[48];              // fixed length in the standard
 *        uint32 verify_result;
 *        opaque peer_cert<0..2^24-1>;    // length 0 means no peer cert
 *        opaque ticket<0..2^24-1>;       // length 0 means no ticket
 *        uint32 ticket_lifetime;
 *        uint8 mfl_code;                 // up to 255 according to standard
 *        uint8 trunc_hmac;               // 0 or 1
 *        uint8 encrypt_then_mac;         // 0 or 1
 *
 *      case MBEDTLS_SSL_MINOR_VERSION_4: // TLS 1.3
 *        uint64 start_time;
 *        uint8 ciphersuite[2];
 *        uint32 ticket_lifetime;
 *        uint32 ticket_age_add;
 *        uint8 ticket_flags;
 *        opaque resumption_key<0..255>;
 *        opaque ticket<0..2^16-1>;
 *
 *   };
 *  
 * } serialized_session;
 * 
 */
```

As a preparation for this, we should add the `minor_ver` field and the logic around it to `development` now, so when we upstream TLS 1.3 support, we only need to add the new switch branch for TLS 1.3.

Note that this is not an API break because we make no stability promises for session serialization: Sessions that were serialized in different versions or different configs are not compatible.

### PR DESCRIPTION
**Pull request by hanno-arm**
Instances of `mbedtls_ssl_session` represent data enabling session resumption.

With the introduction of TLS 1.3, the format of this data changes. We therefore
need TLS-version field as part of `mbedtlsl_ssl_session` which allows distinguish
1.2 and 1.3 sessions.

This commit introduces such a TLS-version field to mbedtls_ssl_session.

The change has a few ramifications:

- Session serialization/deserialization routines need to be adjusted.

  This is achieved by adding the TLS-version after the header of
  Mbed TLS version+config, and by having the subsequent structure
  of the serialized data depend on the value of this field.

  The details are described in terms of the RFC 8446 presentation language.

  The 1.2 session (de)serialization are moved into static helper functions,
  while the top-level session (de)serialization only parses the Mbed TLS
  version+config header and the TLS-version field, and dispatches according
  to the found version.

  This way, it will be easy to add support for TLS 1.3 sessions in the future.

- Tests for session serialization need to be adjusted

- Once we add support for TLS 1.3, with runtime negotiation of 1.2 vs. 1.3,
  we will need to have some logic comparing the TLS version of the proposed session
  to the negotiated TLS version. For now, however, we only support TLS 1.2,
  and no such logic is needed. Instead, we just store the TLS version in the
  session structure at the same point when we populate mbedtls_ssl_context.minor_ver.

The change introduces some overlap between `mbedtls_ssl_session.minor_ver` and
`mbedtls_ssl_context.minor_ver`, which should be studied and potentially resolved.
However, with both fields being private and explicitly marked so, this can happen
in a later change.

Fixes #4802

### CODE REVIEW

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks mostly good, except I don't understand why the context and session structures now both have a minor version field. It's no big deal right now since the library only supports a single version, but if we go down that route adding a version field is pointless. When the library starts supporting multiple version, we'll have to do extra work keeping those fields synchronized.

File: library/ssl_tls.c [hunk 1]
```diff
@@ -4550,44 +4550,62 @@ static unsigned char ssl_serialized_session_header[] = {
...
  * (in the presentation language of TLS, RFC 8446 section 3)
  *
- *  opaque mbedtls_version[3];   // major, minor, patch
- *  opaque session_format[2];    // version-specific 16-bit field determining
- *                               // the format of the remaining
- *                               // serialized data.
+ *  struct {
+ *
+ *    opaque mbedtls_version[3];   // major, minor, patch
+ *    opaque session_format[2];    // version-specific 16-bit field determining
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/ssl_tls.c:4556]: 1. Good point, is shoud library-version-specific.
2. It is mostly for the code-structure: `mbedtls_version` and `session_format` are determined at compile-time, so we can parse and validate it with a simple `memcpy`. Protocol-specific fields follow. I considered integrating the protocol-version here, but found it preferable to keep the structure as we have it now.

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -927,6 +927,8 @@ struct mbedtls_ssl_session
 
     unsigned char exported;
 
+    unsigned char MBEDTLS_PRIVATE(minor_ver);    /*!< The TLS version used in the session. */
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:930]: To keep the change small and allow us to move forwards - that's the only reason. I think a good time to revisit this is #4823.

File: library/ssl_tls.c [hunk 3]
```diff
@@ -4550,44 +4550,62 @@ static unsigned char ssl_serialized_session_header[] = {
...
+ *    uint8 compression;              // 0 or 1
+ *    uint8 session_id_len;           // at most 32
+ *    opaque session_id[32];
+ *    opaque master[48];              // fixed length in the standard
+ *    uint32 verify_result;
+ *    opaque peer_cert<0..2^24-1>;    // length 0 means no peer cert
+ *    opaque ticket<0..2^24-1>;       // length 0 means no ticket
+ *    uint32 ticket_lifetime;
+ *    uint8 mfl_code;                 // up to 255 according to standard
+ *    uint8 trunc_hmac;               // 0 or 1
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/ssl_tls.c:4597]: Sure!

[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/ssl_tls.c]: Fixed.

File: library/ssl_tls.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/ssl_tls.c]: Fixed 1.

**Comment by hanno-arm** [comment]
Thank you @gilles-peskine-arm for your review! I have addressed your comments, and tried to explain why I prefer not to try to consolidate `mbedtls_ssl_context.minor_ver` and `mbedtls_ssl_session.minor_ver` at this point.

File: include/mbedtls/ssl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:930]: Ok, but please:
* Add a comment to the documentation of both fields mentioning the temporary duplication.
* Record the duplication as technical debt to resolve, either in #4823 or a dedicated issue.

[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:930]: Done.

File: library/ssl_tls.c [hunk 4]
```diff
@@ -4560,7 +4560,7 @@ static unsigned char ssl_serialized_session_header[] = {
  *                                 // configuration options which influence
  *                                 // the structure of mbedtls_ssl_session.
  *
- *    uint8_t minor_ver;           // Possible values:
+ *    uint8_t minor_ver;           // Protol-version. Possible values:
```
Comment (mpg) [inline library/ssl_tls.c:4563]: Typo: protocol

**Review by mpg — CHANGES_REQUESTED** [review]
Looks good to me except for a typo in a comment.

File: library/ssl_tls.c (hunk 4 above)
Comment (hanno-arm) [reply library/ssl_tls.c:4563]: Fixed!

**Review by mpg — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]

**Pull request merged**